File: app/src/main/java/com/huanchengfly/tieba/post/adapters/SearchForumAdapter.java
package com.huanchengfly.tieba.post.adapters;

import android.app.Activity;
import android.content.Context;

import com.huanchengfly.tieba.api.models.SearchForumBean;
import com.huanchengfly.tieba.post.R;
import com.huanchengfly.tieba.post.utils.ImageUtil;
import com.huanchengfly.tieba.post.utils.NavigationHelper;
import com.othershe.baseadapter.ViewHolder;
import com.othershe.baseadapter.base.MultiBaseAdapter;

import java.util.ArrayList;
import java.util.List;

public class SearchForumAdapter extends MultiBaseAdapter<SearchForumBean.ForumInfoBean> {
    public static final int TYPE_EXACT = 0;
    public static final int TYPE_FUZZY = 1;
    private NavigationHelper navigationHelper;

    public SearchForumAdapter(Context context) {
        super(context, null, true);
        navigationHelper = NavigationHelper.newInstance(context);
    }

    public void setData(SearchForumBean.DataBean data) {
        List<SearchForumBean.ForumInfoBean> forumInfoBeans = new ArrayList<>();
        if (data.getExactMatch() != null && data.getExactMatch().getForumNameShow() != null) {
            forumInfoBeans.add(data.getExactMatch());
        }
        forumInfoBeans.addAll(data.getFuzzyMatch());
        setNewData(forumInfoBeans);
    }

    private boolean canLoadGlide() {
        if (mContext instanceof Activity) {
            return !((Activity) mContext).isDestroyed();
        }
        return false;
    }

    @Override
    protected void convert(ViewHolder viewHolder, SearchForumBean.ForumInfoBean forumInfoBean, int position, int type) {
        viewHolder.setText(R.id.item_search_forum_title, forumInfoBean.getForumNameShow() + "吧");
        viewHolder.setOnClickListener(R.id.item_search_forum, (view) -> {
            navigationHelper.navigationByData(NavigationHelper.ACTION_FORUM, forumInfoBean.getForumName());
        });
        ImageUtil.load(viewHolder.getView(R.id.item_search_forum_avatar), ImageUtil.LOAD_TYPE_AVATAR, forumInfoBean.getAvatar());
        if (type == TYPE_EXACT) {
            SearchForumBean.ExactForumInfoBean exactForumInfoBean = (SearchForumBean.ExactForumInfoBean) forumInfoBean;
            viewHolder.setText(R.id.item_search_forum_subtitle, exactForumInfoBean.getSlogan());
        }
    }

    @Override
    protected int getItemLayoutId(int type) {
        if (type == TYPE_EXACT) {
            return R.layout.item_search_forum_exact;
        }
        return R.layout.item_search_forum;
    }

    @Override
    protected int getViewType(int i, SearchForumBean.ForumInfoBean forumInfoBean) {
        if (forumInfoBean instanceof SearchForumBean.ExactForumInfoBean) {
            return TYPE_EXACT;
        }
        return TYPE_FUZZY;
    }
}

File: app/src/main/java/com/huanchengfly/tieba/post/adapters/SearchForumAdapter.kt
package com.huanchengfly.tieba.post.adapters

import android.content.Context
import com.alibaba.android.vlayout.layout.LinearLayoutHelper
import com.huanchengfly.tieba.post.R
import com.huanchengfly.tieba.post.adapters.base.BaseMultiTypeDelegateAdapter
import com.huanchengfly.tieba.post.api.models.SearchForumBean
import com.huanchengfly.tieba.post.api.models.SearchForumBean.ExactForumInfoBean
import com.huanchengfly.tieba.post.components.MyViewHolder
import com.huanchengfly.tieba.post.utils.ImageUtil
import com.huanchengfly.tieba.post.utils.getIntermixedColorBackground

class SearchForumAdapter(context: Context?) : BaseMultiTypeDelegateAdapter<SearchForumBean.ForumInfoBean>(context!!, LinearLayoutHelper()) {
    override fun convert(viewHolder: MyViewHolder, item: SearchForumBean.ForumInfoBean, position: Int, viewType: Int) {
        viewHolder.setText(R.id.item_search_forum_title, context.getString(R.string.title_forum, item.forumNameShow))
        ImageUtil.load(viewHolder.getView(R.id.item_search_forum_avatar), ImageUtil.LOAD_TYPE_AVATAR, item.avatar)
        if (viewType == TYPE_EXACT) {
            val exactForumInfoBean = item as ExactForumInfoBean
            viewHolder.setText(R.id.item_search_forum_subtitle, exactForumInfoBean.slogan)
        }
        viewHolder.itemView.background = getIntermixedColorBackground(
                context,
                position,
                itemCount,
                positionOffset = 1,
                colors = intArrayOf(R.color.default_color_card, R.color.default_color_divider),
                radius = context.resources.getDimension(R.dimen.card_radius)
        )
    }

    override fun getItemLayoutId(itemType: Int): Int {
        return if (itemType == TYPE_EXACT) {
            R.layout.item_search_forum_exact
        } else R.layout.item_search_forum
    }

    override fun getViewType(position: Int, item: SearchForumBean.ForumInfoBean): Int {
        return if (item is ExactForumInfoBean) {
            TYPE_EXACT
        } else TYPE_FUZZY
    }

    companion object {
        const val TYPE_EXACT = 0
        const val TYPE_FUZZY = 1
    }
}